freqtrade/strategy/hyper.py

```diff
@@ -0,0 +1,170 @@
+"""
+IHyperStrategy interface, hyperoptable Parameter class.
+This module defines a base class for auto-hyperoptable strategies.
+"""
+from typing import Iterator, Tuple, Any, Optional, Sequence, Union
+
+from skopt.space import Integer, Real, Categorical
+
+from freqtrade.exceptions import OperationalException
+
+
+class BaseParameter(object):
+    """
+    Defines a parameter that can be optimized by hyperopt.
+    """
+    category: Optional[str]
+    default: Any
+    value: Any
+    space: Sequence[Any]
+
+    def __init__(self, *, space: Sequence[Any], default: Any, category: Optional[str] = None,
+                 **kwargs):
+        """
+        Initialize hyperopt-optimizable parameter.
+        :param category: A parameter category. Can be 'buy' or 'sell'. This parameter is optional if
+         parameter field
+         name is prefixed with 'buy_' or 'sell_'.
+        :param kwargs: Extra parameters to skopt.space.(Integer|Real|Categorical).
+        """
+        if 'name' in kwargs:
+            raise OperationalException(
+                'Name is determined by parameter field name and can not be specified manually.')
+        self.category = category
+        self._space_params = kwargs
+        self.value = default
+        self.space = space
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}({self.value})'
+
+    def get_space(self, name: str) -> Union[Integer, Real, Categorical]:
+        raise NotImplementedError()
+
+
+class IntParameter(BaseParameter):
+    default: int
+    value: int
+    space: Sequence[int]
+
+    def __init__(self, *, space: Sequence[int], default: int, category: Optional[str] = None,
+                 **kwargs):
+        """
+        Initialize hyperopt-optimizable parameter.
+        :param space: Optimization space, [min, max].
+        :param default: A default value.
+        :param category: A parameter category. Can be 'buy' or 'sell'. This parameter is optional if
+         parameter field
```

Comment:
The more i look at the `space` and `category` options, the more it confuses me (from a naming perspective only)

I understand the `space` is the space i wanna optimize (min / max value), and category would be buy / sell.
The confusing part is with hyperopt - where we have `--spaces buy sell roi` - to optimize only certain parts - which are called spaces in the cli parameter - but suddenly corresponds to a "category" in this context.

I guess technically both are spaces (one is the space within this parameter - e.g. [-20, 20] ) - while the other is the overall (bigger scoped) search space of buy / sell ... 


---

The problem is - i don't have a better alternative at the moment (and i suspect that's how you got to these names in the first place too :laughing: )

Reply:
You are right. Another point i am not entirely happy with `space` is that it is kind of bit loose. You could pass a list with three items for int/float and get a runtime exception. skopt avoids this by explicitly passing min and max parameters. If we did that for int/float, we could use "choices" for categorical, that would free up "space" name to use in place of "category". We would have a consistency of terminology. Although i do admit i like `whatever_name=[0, 100]` better than `min=0, max=100`.

Reply:
hm, maybe we can adapt the "skopt" approach - using positional arguments (`IntParameter(0, 20, default=10, spaces=['buy'])`)?

it might not be as expressive than using low/high (that's the naming skopt uses for Int/Real spaces).
potentially we can even have both ways?

``` python
x = IntParameter(low=0, high=20, default=10, spaces=['buy'])
y = IntParameter([0, 20], default=10, spaces=['buy'])
```

with the 2nd version exploding the first parameter to low + high ....

Reply:
Sounds like a perfect compromise. I love that `spaces=[]` touch. Would probably be useful for defining same categorial parameter for both spaces _once_.

Reply:
jup ... it's actually ... not intentional that i changed it from how category works :D 

I'm however not 100% sure if it will work with more than one space if i think about it.

hyperopt would change the SAME parameter from 2 different generators ... but there's only one value / variable in the strategy - so you can't know which one was used (the output would have it in both spaces - but you can't be sure which one was actually used as the 2 would overwrite each other).

Reply:
Oh good point. Lets ignore that then.. I mean we could make it work, but marginal usefulness it totally not worth extra complexity